text/0007-performance-data-propagation.md

```diff
@@ -0,0 +1,179 @@
+            {
+                var transportProfilerData = profileTransport.GetTransportGetData();
+                transportProfilerData.ForEach(x =>
+                {
+                    if (!profilerData.tickData.ContainsKey(x.Key)) profilerData.tickData.Add(x.Key, x.Value);
+                });
+            }
+
+            PerfomanceDataEvent?.Invoke(profilerData);
+        }
+    }
+}
+
+
+namespace MLAPI.Transport
+{
+    class UNetTransport : Transport, ITransportProfilerData
+    {
+        private static Dictionary<string, int> transportProfilerData;
+        private static bool profilerEnabled;
+
+        void Init()
+        {
+            transportProfilerData = new Dictionary<string, int>();
+        }
+
+        Dictionary<string, int> GetTransportGetData()
+        {
+            return transportProfilerData;
+        }
+
+        void Send(ulong clientId, ArraySegment<byte> data, string channelName)
+        {
+            if (profilerEnabled)
+            {
+                transportProfilerData[ProfilerConstants.NumberOfBuffMessagesOut] += 1;
+            }
+        }
+    }
+}
+
+namespace ThirdParty
+{
+    class StatsRenderer
+    {
+        void Start()
+        {
+            NetworkManager.instance.PerfomanceDataEvent += OnPerformanceEvent;
+        }
+
+        void OnPerformanceEvent(ProfilerTickData profilerData)
+        {
+            IMGUI.textlabel("Num Server RPCs", profilerData.tickData[ProfilerConstants.NumberOfServerRPCs]);
+        }
+    }
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+With a dictionary we occur the performance overhead of boxing somewhat. Additionally we require our callers to look up the appropriate types for the data that they are pulling out of the dictionary.
+
+However, changes to the `ProfilerTickData` will not cause large scale changes to the transport, MLAPI, and all third party consumers.
+
+Each transport is also required to implement any required changes needed to fill out the relevant transport-level information. Of course not filling out the fields simply means the Overlay, Profiler, Logger will just display default values for the field types. This means they can implement after an initial release of their transport, if need be.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This design of single memory block is the best for locality of data.
+
+There is less room for user error since there is only one readonly object that they get access to.
+
+There is no room for user injected data or exceptions since we aren’t using an event signaling or actions. All user interaction for rendering happens at end of tick so they can’t affect the time of profiling or cause unforeseen issues.
+
+The impact of not doing a Network Tick and Profiler Notification System is that we’d loose the Profiling capabilities that existed in the vanilla Profiler with Mirror/UNET.
+
+# Prior art
+[prior-art]: #prior-art
+
```

Comment:
There is already a stat gathering feature added to MLAPI that might be extendable / have things we can steal (see ProfilerStat / ProfilerStatManager).  It does a moving average.  In particular I had to put in some smarts in the 'Record' method to handle when tons of samples are taken in a short period of time.

Reply:
That code will still be used. This should feed into this for the Profiler among other things.